File: nodejs/src/Tree/InvertBinaryTree/InvertBinaryTree.ts
import BaseTreeAlgorithmT from "../BaseTreeAlgorithm";
import TreeNodeT from "../TreeNode";

class InvertBinaryTree<T> extends BaseTreeAlgorithmT<T>{

  exec(node: TreeNodeT<T> | null): TreeNodeT<T> | null{
    // handle special case first
    if(node === null){
      return null;
    }

    // invert for current node
    const temp: TreeNodeT<T> | null = node.left;
    node.left = node.right;
    node.right = temp;

    // recursively invert left and right node
    this.exec(node.left)
    this.exec(node.right)

    // return the result
    return node;
  }
}

export default InvertBinaryTree

File: go/src/tree/invert_binary_tree/invert_binary_tree.go
package invert_binary_tree

import "algorithms/go/tree"

func invertTree(root *tree.TreeNode) *tree.TreeNode {
	// base case
	if root == nil {
		return nil
	}

	// swap left & right node
	root.Left, root.Right = root.Right, root.Left

	// invert left & right node
	invertTree(root.Left)
	invertTree(root.Right)

	return root
}
